### ISSUE TITLE
Several Modules incompatible with LifecycleOperation & patches to make them compatible

### ISSUE THREAD

**Issue by edlongman** [inet-framework/inet#527]
Several problems here but bear with me as I think some have simple solutions.

Setup 
----
`DymoRouter` with `Ieee802154NarrowbandMac` wireless interface. 
`SimpleEpEnergyStorage` and `SimpleEpEnergyManagement` operating a threshold on off system based on energy received from `AlternatingEpEnergyGenerator`

`wlan[0].radio` has `SensorStateBasedEpEnergyConsumer`

Network of 10 nodes with `UdpBasicApp` and `UdpSink` simulating "load" on network.

Problems and Patches
-------

* `Radio.cc` implements `handleStopOperation` and deletes any in progress transmissions on shutdown. `Ieee802154Mac` and presumably other MAC modules subsequently causes an unhandled null pointer Exception when accessing properties of `currentTxFrame`
https://github.com/inet-framework/inet/blob/76f495bca447551d7528c0dd5c153ecdebf94147/src/inet/linklayer/ieee802154/Ieee802154Mac.cc#L450-L455
`if (event == EV_FRAME_TRANSMITTED && currentTxFrame != nullptr)` 
Is enough to check that the currentTxFrame hasn't been deleted

* The `OperationalBase` picks up uncancelled selfMessages like `ccaTimer` and `rxAckTimer` after stopping. This throws a RuntimeError. 
How this should be fixed is up for discussion. Should the MacLayer implement `STOPPING_OPERATION` where it finishes the current transmission but then drops the rest of the queue? In a `CrashOperation` which is what the patch below is, the self messages should be cancelled by the MAC implementation where they are defined.
```
void Ieee802154Mac::handleStopOperation(LifecycleOperation *operation) {
    //TODO: More gracefully allow current Tx to end (delay operation)
    //Cancel all self message timers
    cancelEvent(backoffTimer);
    cancelEvent(ccaTimer);
    cancelEvent(sifsTimer);
    cancelEvent(rxAckTimer);
    MacProtocolBase::handleStopOperation(operation);
}

void Ieee802154Mac::handleCrashOperation(LifecycleOperation *operation) {
    cancelEvent(backoffTimer);
    cancelEvent(ccaTimer);
    cancelEvent(sifsTimer);
    cancelEvent(rxAckTimer);
    MacProtocolBase::handleCrashOperation(operation);
}
```

* `UdpSocket::close()` before created causes RuntimeError https://github.com/inet-framework/inet/blob/76f495bca447551d7528c0dd5c153ecdebf94147/src/inet/transportlayer/contract/udp/UdpSocket.cc#L313 Because gate is uninitiated (nullptr) but `UdpBasicApp::handleStopOperation()` can call `close()` but no check is done to make sure the `socketState` is even `CONNECTED`. Likewise for `UdpSocket::destroy()`.
Patch is to exit early from `UdpSocket::close()` and `UdpSocket::destroy()` is `this->gateToUdp==nullptr`.
```
if(this->getToUdp ==nullptr) return;
```

* `Dymo::handleStopOperation` leaves `RRepTimer` message hanging which is then caught by  `OperationalBase` picking up uncancelled selfMessages. This does raise the question of how this layer should `handleStopOperation` as it could send `RERR` as commented. However since the ipv4 layer is stopped first this would not be possible I believe. 
For now the `cancelRReqTimer(t)` in `cancelRouteDiscovery(target)` to the target fixes this (last line added).
```
void Dymo::cancelRouteDiscovery(const L3Address& target)
{
    EV_INFO << "Canceling route discovery: originator = " << getSelfAddress() << ", target = " << target << endl;
    ASSERT(hasOngoingRouteDiscovery(target));
    auto lt = targetAddressToDelayedPackets.lower_bound(target);
    auto ut = targetAddressToDelayedPackets.upper_bound(target);
    for (auto it = lt; it != ut; it++)
        dropDelayedDatagram(it->second);
    eraseDelayedDatagrams(target);
    cancelRreqTimer(target);
}
```
I can make a pull request for the part of this that you think should become permenant fixes.

**Comment by levy** [comment]
In general, the idea is that all active modules must implement the stop/crash lifecycle operations. They must cancel their timers and stop their processing to prevent sending messages to passive modules after the lifecycle operation is completed. In general, the lifecycle operation API supports extending to a non-zero simulation time, so modules can gracefully close connections, notify peers, etc.

Unfortunately, some INET modules do not implement the lifecycle operations, but this is clearly a problem in these modules. If you look at other MAC implementations, Ethernet for example, they actually implement the lifecycle support one way or another.

In the UdpBasicApp, I think the socket must not be closed if it hasn't been opened in the first place.

As for the Dymo routing protocol, it could in principle send the RERR but only if the Ipv4 module and all the underlying modules support a non-zero simulation time stop operation. Otherwise, you're right that the timers should be canceled. I'm not sure if cancelRreqTimer should be called from cancelRouteDiscovery though, because I'm not sure that is it the case that it should always be called (e.g. during normal operation)? If it's only required during stop/crash, then this call should be in the handlers of those operations.

Thanks for reporting these issues and the offer to send a pull requres. I would prefer to have a pull request and continue the discussions there to have the reasoning behind the changes for future reference. Also, if it's actually you who does the development and testing, then I guess you should be the author in the patches. At least we will have more contributors and you can take the credit for your contributions.

**Comment by edlongman** [comment]
Is it possible for a high level module to check if lower level modules implements extended shutdown time other than navigating through the structure and checking `NodeStatus::GOING_DOWN` or something similar?

Shall I split into 3 pull requests: (Mac and Phy ), (Dymo and ipv4), (UdpSocket and UdpApp)?

**Comment by edlongman** [comment]
Continuing to record the debuging story here:

`Message received when Ipv4 is down` is because `Dymo` uses `sendDelayed` message primitive. This is discouraged by the [manual (sec: delayed sending)](https://doc.omnetpp.org/omnetpp/manual/#sec:simple-modules:delayed-sending). 
`sendDelayed` is used the implementation of the `jitter` parameter of the `Dymo` module. 

Quick Fix: Set the jitter to 0.

Ultimate Fix: Implement Queue of `RReq` responses via a `PacketDelayer`. On `shutdownOperation` clear the `RReq` delayer.

Instead a queue of self messages should be

**Comment by edlongman** [comment]
@levy Is there something more that needs to be done for the associated pull requests to get them merged?

**Comment by levy** [comment]
I'm sorry, I can't take a look at it now, because I'm on holiday this week.

**Comment by levy** [comment]
Thank you for your efforts! My colleague, @ZoltanBojthe will take care of merging your pull requests.

In any case, we usually fix most bugs and merge most pull request before releasing a new version, so these PRs will be merged in version 4.3 at the latest.
